unseal.go

```diff
@@ -0,0 +1,159 @@
+// will be returned, and the TPM's dictionary attack counter will be incremented.
+//
+// If the TPM's dictionary attack logic has been triggered, a ErrTPMLockout error will be returned.
+//
+// If the TPM is not provisioned correctly, then a ErrTPMProvisioning error will be returned. In this case, ProvisionTPM should be
+// called to attempt to resolve this.
+//
+// If the TPM sealed object cannot be loaded in to the TPM for reasons other than the lack of a storage root key, then a
+// InvalidKeyFileError error will be returned. This could be caused because the sealed object data is invalid in some way, or because
+// the sealed object is associated with another TPM owner (the TPM has been cleared since the sealed key data file was created with
+// SealKeyToTPM), or because the TPM object at the persistent handle reserved for the storage root key has a public area that looks
+// like a valid storage root key but it was created with the wrong template. This latter case is really caused by an incorrectly
+// provisioned TPM, but it isn't possible to detect this. A subsequent call to SealKeyToTPM or ProvisionTPM will rectify this.
+//
+// If the TPM's current PCR values are not consistent with the PCR protection policy for this key file, a InvalidKeyFileError error
+// will be returned.
+//
+// If any of the metadata in this key file is invalid, a InvalidKeyFileError error will be returned.
+//
+// If the TPM is missing any persistent resources associated with this key file, then a InvalidKeyFileError error will be returned.
+//
+// If the key file has been superceded (eg, by a call to UpdateKeyPCRProtectionPolicy), then a InvalidKeyFileError error will be
+// returned.
+//
+// If the signature of the updatable part of the key file's authorization policy is invalid, then a InvalidKeyFileError error will
+// be returned.
+//
+// If the metadata for the updatable part of the key file's authorization policy is not consistent with the approved policy, then a
+// InvalidKeyFileError error will be returned.
+//
+// If the provided PIN is incorrect, then a ErrPinFail error will be returned and the TPM's dictionary attack counter will be
+// incremented.
+//
+// If access to sealed key objects created by this package is disallowed until the next TPM reset or TPM restart, then a
+// ErrSealedKeyAccessLocked error will be returned.
+//
+// If the authorization policy check fails during unsealing, then a InvalidKeyFileError error will be returned. Note that this
+// condition can also occur as the result of an incorrectly provisioned TPM, which will be detected during a subsequent call to
+// SealKeyToTPM.
+//
+// On success, the unsealed cleartext key is returned. If the lock argument is true, then subsequent access to all sealed key
+// objects created by this package will be denied until the next TPM reset or TPM restart.
+func (k *SealedKeyObject) UnsealFromTPM(tpm *TPMConnection, pin string, lock bool) ([]byte, error) {
```

Comment:
will this typically be used with lock=true?

pin and lock could go in some kind of UnsealOptions struct

Reply:
Yes, I imagine that it will be used with lock=true. There will also need to be a separate API for triggering the locking without unsealing too.

Reply:
Following the conversation on IRC, I've removed the lock option from this in favour of a separate API (which will come in a future PR).